pkg/webhook/admission/pod/storage_initializer_injector.go

```diff
@@ -425,16 +425,124 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
 		}
 	}
 
+	// Add init container to the spec
+	pod.Spec.InitContainers = append(pod.Spec.InitContainers, *initContainer)
+
+	return nil
+}
+
+// SetIstioCniSecurityContext determines if Istio is installed in using the CNI plugin. If so,
+// the UserID of the storage initializer is changed to match the UserID of the Istio sidecar.
+// This is to ensure that the storage initializer can access the network.
+func (mi *StorageInitializerInjector) SetIstioCniSecurityContext(pod *v1.Pod) error {
+	// Find storage initializer container
+	var storageInitializerContainer *v1.Container
+	for idx, c := range pod.Spec.InitContainers {
+		if c.Name == StorageInitializerContainerName {
+			storageInitializerContainer = &pod.Spec.InitContainers[idx]
+		}
+	}
+
+	// If the storage initializer is not injected, there is no action to do
+	if storageInitializerContainer == nil {
+		return nil
+	}
+
 	// Allow to override the uid for the case where ISTIO CNI with DNS proxy is enabled
 	// See for more: https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers.
 	if value, ok := pod.GetAnnotations()[constants.IstioSidecarUIDAnnotationKey]; ok {
 		if uid, err := strconv.ParseInt(value, 10, 64); err == nil {
-			initContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
+			if storageInitializerContainer.SecurityContext == nil {
+				storageInitializerContainer.SecurityContext = &v1.SecurityContext{}
+			}
+			storageInitializerContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
+		}
+	} else {
+		// When Istio CNI is enabled, a sidecar.istio.io/interceptionMode annotation
+		// is injected to the pod.
+		istioInterceptionMode, istioCniEnabled := pod.Annotations[constants.IstioInterceptionModeAnnotation]
```

Comment:
Is there some other way to detect Istio integration eg. config?

Reply:
Also would `k8s.v1.cni.cncf.io/network-status` be a better option here?

Reply:
I don't see this annotation or label in `minikube`

Reply:
> Is there some other way to detect Istio integration eg. config?

I compared the install YAML with istio-cni enabled and disabled. The most relevant thing I'm able to see is the presence of an `istio-cni` ConfigMap in istio namespace. Not sure if that is really better.

FWIW, I chose this annotation based on this Istio source: https://github.com/istio/istio/blob/f5b75167e06b601ffe50a6e9b5ee2ada10b0c2cf/manifests/charts/istiod-remote/files/injection-template.yaml#L48-L52. The annotations is always there if CNI is enabled.

In a line before it, there is the `k8s.v1.cni.cncf.io/networks` which is related, but that one has a conditional...

Reply:
Regarding `k8s.v1.cni.cncf.io/network-status` I was looking at OCP but it seems OCP does not support chained mode so thus the annotation.

>  The annotations is always there if CNI is enabled.

If the annotation is there CNI is on always?

I also found this, I think that should work given that we check the pod at the webhook side:
https://github.com/istio/istio/blob/26cf55f5d6f576bf8fad436b61c7ec1240fb671d/pkg/config/analysis/analyzers/maturity/maturity.go#L134. 

So with cni instead of the istio-init container, you should get an [istio-validation] one (https://istio.io/latest/docs/setup/additional-setup/cni/#race-condition--mitigation). Verified it on minikube and [OCP](https://docs.openshift.com/container-platform/4.14/service_mesh/v2x/ossm-vs-community.html#ossm-cni_ossm-vs-istio). On OCP btw there is no init container injected.

Reply:
> If the annotation is there, is CNI always on?
> 

From the template, yes, that's what I understand.

> I also found this, I think that should work given that we check the pod at the webhook side:
> 
> https://github.com/istio/istio/blob/26cf55f5d6f576bf8fad436b61c7ec1240fb671d/pkg/config/analysis/analyzers/maturity/maturity.go#L134. 
> 
> So with cni instead of the istio-init container, you should get an [istio-validation](https://istio.io/latest/docs/setup/additional-setup/cni/#race-condition--mitigation) one. Verified it on minikube and [OCP](https://docs.openshift.com/container-platform/4.14/service_mesh/v2x/ossm-vs-community.html#ossm-cni_ossm-vs-istio) that with cni you get no istio-init container. On OCP btw there is no init container injected at all.
> 

I'll check. If it works in both cases, I can change it to align to this code. Yet, as noted in the comments, the check to the interception mode is still needed (to rule out Ambient mode).

Reply:
Unfortunately, the `istio-validation` container is missing in Maistra.
So, current approach of checking the interception mode annotations seems to be the most reliable way.

Reply:
Oh... Although after re-reading the suggested code, it checks for presence of `istio-init`... let me re-test.

Reply:
@skonto OK, I changed the code to use your suggested approach.